honeybee_3dm/grid.py

```diff
@@ -0,0 +1,77 @@
+"""Import Rhino surfaces as Honeybee grid objects from a Rhino3DM file."""
+import warnings
+# The Rhino3dm library provides the ability to access content of a Rhino3dm
+# file from outside of Rhino
+import rhino3dm
+
+# Importing Honeybee dependencies
+from honeybee_radiance.sensorgrid import SensorGrid
+from honeybee.typing import clean_and_id_string, clean_string
+from ladybug_geometry.geometry3d import Face3D, Mesh3D
+
+# Importing dependencies from Honeybee-3dm package
+from .togeometry import brep2d_to_face3d, mesh_to_face3d, mesh_to_mesh3d
+from .helper import filter_objects_by_layer
+
+
+# TODO: expose an option to change the target layer name
+# TODO: Grid size and offset will have to provided using the config file
+
+
+def import_grids(rhino3dm_file, tolerance, grid_size=1, grid_offset=0):
+    """Creates Honeybee grids from a rhino3dm file.
+
+    This function assumes all the grid objects are under a layer named ``grid``.
+
+    Args:
+        rhino3dm_file: The rhino file from which Honeybee grids will be created.
+        tolerance: A rhino3dm tolerance object. Tolerance set in the rhino file.
+        grid_size: Grid size. Defaults to 1.
+        grid_offset: Grid offset from the grid geometry. Defaults to 0.
+
+    Returns:
+        A list of Honeybee grids.
+    """
+
+    # Objects from rhino file
+    try:
+        grid_objs = filter_objects_by_layer(rhino3dm_file, 'grid')
+        warnings.warn(
+            'Only objects on layer "grid" will be used to create grids.')
+    except ValueError:
+        # no layer named grid
+        return []
+
+    hb_grids = []
+
+    for obj in grid_objs:
+        try:
+            geo = obj.Geometry
+            # If it's a Brep
+            if isinstance(geo, rhino3dm.Brep):
+                face3d = brep2d_to_face3d(geo, tolerance)[0]
+                mesh3d = face3d.mesh_grid(grid_size, grid_size, grid_offset)
+
+            # If it's a Mesh
+            elif isinstance(geo, rhino3dm.Mesh):
+                # If it's a Mesh with only one face
+                if len(geo.Faces) == 1:
+                    face3d = mesh_to_face3d(geo)[0]
+                    mesh3d = face3d.mesh_grid(grid_size, grid_size, grid_offset)
+                # It it's a mesh with multiple faces
+                elif len(geo.Faces) > 1:
+                    mesh3d = mesh_to_mesh3d(geo)
+        except AssertionError:
+            warnings.warn(
+                'Grids are not created for Breps with curved edges.'
+                ' Please mesh them in Rhino and try again.')
+            # TODO: Find a way to Implement a method to create grids from a planar
+            # TODO geometry with curved edges
+            pass
```

Comment:
What if we use `GetMesh` method to get the render mesh in Rhino? Here is an example:

https://github.com/jesterKing/import_3dm/blob/master/import_3dm/converters/render_mesh.py#L52-L57

Reply:
We're using the `Any` object to create mesh. Please see [here](https://github.com/devngc/honeybee-3dm/blob/master/honeybee_3dm/togeometry.py#L140)
It's not we are not able to deal with a planar brep that has curved edges. We are and we're doing that through meshing using the `Any` object similar to jesterKing. That mesh is traingulated mesh and not very useful to create grid points in my opinion. At the bottom of the image below, we have a mesh created from Rhino3dm. At the top of the image, we have a mesh created in rhino and then used for grids. I believe, you would agree that the top one is what we want and that's why I am asking the user to mesh it in rhino and try again.
![why avoid curved brep](https://user-images.githubusercontent.com/20086924/99156397-8300a080-268e-11eb-9d23-23209fe45c3a.jpg)

You would also ask, "how ladybug-rhino is doing this?" Well, we have methods to use the meshing parameters in rhinocommon and ladybug-rhino is using them.

While we have [meshing parameters](https://mcneel.github.io/rhino3dm/python/api/MeshingParameters.html?highlight=meshingparameter) in Rhino3dm, apparently, there's no direct way to use them. See [here](https://mcneel.github.io/rhino3dm/python/api/BrepFace.html#rhino3dm.BrepFace.GetMesh) also.

I [asked](https://discourse.mcneel.com/t/rhino3dm-meshingparameters-python/112216?u=devang_chauhan) on the forum. I am waiting for Steve Baer's response and if I don't hear from him soon then I am going to ask him why we have this whole class and no way to use it.

Reply:
I agree that the render mesh is not the best mesh for analysis but I think we should still let the study run. You are already providing a warning to user.

I don't think meshing will ever be involved in Rhino3dm. It will be kept as part of RhinoCompute.

That being said, and I have been thinking about this for some time, since this library will run outsider Rhino and IronPython we might be able to use Blender's meshing methods.

Hi @Moult, can you point us to a code sample to generate a QuadMesh using Blender library? Thanks!

Reply:
@mostaphaRoudsari Blender has a ridiculous number of ways to remesh things, some manual, some semi automated, and some fully automated - I don't know the full history behind this but I suspect in Blender because meshes are first-class citizens in my initial implementation I left it to the user to define their topology.

But if you are looking for algorithms, here's a few options to peruse:

 - Remesh modifier https://docs.blender.org/manual/en/latest/modeling/modifiers/generate/remesh.html
 - Quadriflow remesh https://docs.blender.org/manual/en/latest/modeling/meshes/retopology.html
 - Tris to quads https://docs.blender.org/manual/en/latest/modeling/meshes/editing/face/triangles_quads.html (looking at the upper screenshot, tris to quads could further improve on the output)

Reply:
Thanks @Moult, Tris to Quads looks promising. @mostaphaRoudsari, we should definitely use more of Blender.

Reply:
Thanks @Moult for the links. Where should we look for some sample Python code to do something like this:

![image](https://user-images.githubusercontent.com/2915573/99159504-e13b7c80-26aa-11eb-849e-47803dbee6b2.png)

from [here](https://docs.blender.org/manual/en/latest/modeling/modifiers/generate/remesh.html). Sorry if it is a very obvious question for Blender users but I couldn't find a useful code sample by searching `blender python remesh`. Cheers.

Reply:
@mostaphaRoudsari I have a sneaking suspicion it is C++, not Python unfortunately. I'm not familiar with the C++ codebase.

See https://git.blender.org/gitweb/gitweb.cgi/blender.git
This seems like the entry point to begin code diving for the remesh modifier: https://git.blender.org/gitweb/gitweb.cgi/blender.git/blob_plain/HEAD:/source/blender/modifiers/intern/MOD_remesh.c

I think sharp mode is the one you're looking for. There is a Github mirror which makes life easier when searching: https://github.com/blender/blender/search?q=remesh

Reply:
I'm resolving this for now. @devngc, we should probably open an issue and link this conversation to address it at some point in the near future. Thanks @Moult!

Reply:
Created a new issue